clients/src/main/java/org/apache/kafka/common/requests/InitProducerIdRequest.java

```diff
@@ -16,106 +16,71 @@
-        }
+        private final InitProducerIdRequestData data;
 
-        public Builder(String transactionalId, int transactionTimeoutMs) {
+        public Builder(InitProducerIdRequestData data) {
             super(ApiKeys.INIT_PRODUCER_ID);
-
-            if (transactionTimeoutMs <= 0)
-                throw new IllegalArgumentException("transaction timeout value is not positive: " + transactionTimeoutMs);
-
-            if (transactionalId != null && transactionalId.isEmpty())
-                throw new IllegalArgumentException("Must set either a null or a non-empty transactional id.");
-
-            this.transactionalId = transactionalId;
-            this.transactionTimeoutMs = transactionTimeoutMs;
+            this.data = data;
         }
 
         @Override
         public InitProducerIdRequest build(short version) {
-            return new InitProducerIdRequest(version, transactionalId, transactionTimeoutMs);
+            if (data.transactionTimeoutMs() <= 0)
+                throw new IllegalArgumentException("transaction timeout value is not positive: " + data.transactionTimeoutMs());
+
+            if (data.transactionalId() != null && data.transactionalId().isEmpty())
+                throw new IllegalArgumentException("Must set either a null or a non-empty transactional id.");
+
+            return new InitProducerIdRequest(data, version);
         }
 
         @Override
         public String toString() {
-            return "(type=InitProducerIdRequest, transactionalId=" + transactionalId + ", transactionTimeoutMs=" +
-                    transactionTimeoutMs + ")";
+            return data.toString();
         }
     }
 
-    public InitProducerIdRequest(Struct struct, short version) {
-        super(ApiKeys.INIT_PRODUCER_ID, version);
-        this.transactionalId = struct.get(NULLABLE_TRANSACTIONAL_ID);
-        this.transactionTimeoutMs = struct.getInt(TRANSACTION_TIMEOUT_KEY_NAME);
+    public final InitProducerIdRequestData data;
+
+    private InitProducerIdRequest(InitProducerIdRequestData data, short version) {
+        super(ApiKeys.CREATE_TOPICS, version);
+        this.data = data;
     }
 
-    private InitProducerIdRequest(short version, String transactionalId, int transactionTimeoutMs) {
-        super(ApiKeys.INIT_PRODUCER_ID, version);
-        this.transactionalId = transactionalId;
-        this.transactionTimeoutMs = transactionTimeoutMs;
+    public InitProducerIdRequest(Struct struct, short version) {
+        super(ApiKeys.CREATE_TOPICS, version);
+        this.data = new InitProducerIdRequestData(struct, version);
     }
 
+
     @Override
     public AbstractResponse getErrorResponse(int throttleTimeMs, Throwable e) {
-        return new InitProducerIdResponse(throttleTimeMs, Errors.forException(e));
+        InitProducerIdResponseData response = new InitProducerIdResponseData()
+                .setErrorCode(Errors.forException(e).code())
+                .setProducerId(RecordBatch.NO_PRODUCER_ID)
```

Comment:
Probably makes sense to set -1 as the default for `ProducerId` in `InitProducerIdResponse`

Same for `ProducerEpoch`